### PR TITLE
Fb asim 2859 state variables plugin functions

### PR DESCRIPTION
**Pull request by tarcisiofischer**

### CODE REVIEW

File: alfasim_sdk/hook_specs.py [hunk 1]
```diff
@@ -208,6 +208,94 @@ def calculate_entrained_liquid_fraction(
...
+    1) In the plugin python configuration file, define the
+       alfasim_get_phase_properties_calculated_from_plugin function to set which phases
+       the current plugin is able to calculate state variables.
+
+    2) The plugin must, then, implement three hooks:
+    - HOOK_INITIALIZE_STATE_VARIABLE_CALCULATOR
+    - HOOK_CALCULATE_STATE_VARIABLE
+    - HOOK_FINALIZE_STATE_VARIABLE_CALCULATOR
+
+    The first and last hooks are called immediately before the state variables are calculated.
```
Comment (ggrbill) [inline alfasim_sdk/hook_specs.py]: the `HOOK_FINALIZE_STATE_VARIABLE_CALCULATOR` is called _before_ the state variables are calculated?

File: alfasim_sdk/hook_specs.py [hunk 2]
```diff
@@ -208,6 +208,94 @@ def calculate_entrained_liquid_fraction(
...
+    To define a function that'll calculate the state variables for a given phase on ALFAsim,
+    two main steps must be performed:
+
+    1) In the plugin python configuration file, define the
+       alfasim_get_phase_properties_calculated_from_plugin function to set which phases
+       the current plugin is able to calculate state variables.
+
+    2) The plugin must, then, implement three hooks:
+    - HOOK_INITIALIZE_STATE_VARIABLE_CALCULATOR
+    - HOOK_CALCULATE_STATE_VARIABLE
```
Comment (ggrbill) [inline alfasim_sdk/hook_specs.py]: How about `HOOK_CALCULATE_PHASE_PAIR_STATE_VARIABLE`? Should it be listed here?

File: alfasim_sdk/hook_specs.py [hunk 3]
```diff
@@ -208,6 +208,94 @@ def calculate_entrained_liquid_fraction(
...
+    """
+    Hook to calculate the state variable given by the `property_id` (See alfasim_sdk_api common
+    headers to retrieve the available property ids), for the phase `phase_id` (Note that the phase
+    id is the same as the one retrieved from the `get_phase_id()` API function - It is not advisable
+    to use hardcoded numbers).
+
+    The output parameter must be filled with the calculated property for each control volume. The
+    pressure 'P' and layer or mixture temperature 'T' (Depending on the energy model being used)
+    are given in order to perform the calculation. The number of control volumes and the number of
+    layers are also given for convenience.
```
Comment (ggrbill) [inline alfasim_sdk/hook_specs.py]: Number of layers is not given anymore.

File: alfasim_sdk/hook_specs.py [hunk 4]
```diff
@@ -208,6 +208,94 @@ def calculate_entrained_liquid_fraction(
...
+) -> 'int':
+    """
+    Hook to calculate the state variable given by the `property_id` (See alfasim_sdk_api common
+    headers to retrieve the available property ids), for the phase pair `(phase1_id, phase2_id)`
+    (Note that the phase id is the same as the one retrieved from the `get_phase_id()` API function
+    - It is not advisable to use hardcoded numbers).
+
+    The output parameter must be filled with the calculated property for each control volume. The
+    pressure 'P' and mixture temperature 'T_mix' are given in order to perform the calculation.
+    The number of control volumes and the number of layers are also given for convenience.
```
Comment (ggrbill) [inline alfasim_sdk/hook_specs.py]: Number of layers is not given anymore.

File: alfasim_sdk/hook_specs.py [hunk 5]
```diff
@@ -208,6 +208,94 @@ def calculate_entrained_liquid_fraction(
...
+    P: "void*",
+    T_mix: "void*",
+    n_control_volumes: "int",
+    phase1_id: "int",
+    phase2_id: "int",
+    property_id: "int",
+    output: "void*"
+) -> 'int':
+    """
+    Hook to calculate the state variable given by the `property_id` (See alfasim_sdk_api common
```
Comment (ggrbill) [inline alfasim_sdk/hook_specs.py]: Perhaps add a list of properties that are related to a pair of phases.

File: alfasim_sdk/hook_specs_gui.py [hunk 6]
```diff
@@ -17,3 +17,31 @@ def alfasim_get_additional_variables():
...
+
+
+@hookspec
+def alfasim_get_phase_properties_calculated_from_plugin():
+    """
+    Must return a list of phases in which state variables will be computed for. In order to
+    implement the properties, HOOK_CALCULATE_STATE_VARIABLE must be implemented on the C plugin.
+
+    Example:
+    from alfasim_sdk.constants import GAS_PHASE
```
Comment (ggrbill) [inline alfasim_sdk/hook_specs_gui.py]: I think that an user defined phase could also be added as an example, like `solid`.

File: alfasim_sdk_api/common.h [hunk 7]
```diff
@@ -31,6 +31,17 @@ enum TimestepScope
     PREVIOUS=1
 };
 
+enum StateVariable {
```
Comment (ggrbill) [inline alfasim_sdk_api/common.h]: I don't know if here is the correct place, but I think would be nice add a description to each of StateVariables, just to let clear what each one represents.

File: alfasim_sdk/hook_specs.py (hunk 1 above)
Comment (tarcisiofischer) [reply alfasim_sdk/hook_specs.py]: after

File: alfasim_sdk/hook_specs.py (hunk 2 above)
Comment (tarcisiofischer) [reply alfasim_sdk/hook_specs.py]: yes

File: alfasim_sdk/hook_specs.py (hunk 3 above)
Comment (tarcisiofischer) [reply alfasim_sdk/hook_specs.py]: removed

File: alfasim_sdk/hook_specs.py (hunk 4 above)
Comment (tarcisiofischer) [reply alfasim_sdk/hook_specs.py]: removed

File: alfasim_sdk/hook_specs.py (hunk 5 above)
Comment (tarcisiofischer) [reply alfasim_sdk/hook_specs.py]: done

File: alfasim_sdk/hook_specs_gui.py (hunk 6 above)
Comment (tarcisiofischer) [reply alfasim_sdk/hook_specs_gui.py]: done

File: alfasim_sdk_api/common.h (hunk 7 above)
Comment (tarcisiofischer) [reply alfasim_sdk_api/common.h]: done

**Pull request merged**